File: python/claude_trace/_session.py
"""AgentSession: context manager tracking a complete agent conversation.

Records cumulative token counts and cost, and creates an OTel root span
that all turn spans will be attached to as children.

Usage::

    with claude_trace.session("billing-agent", customer_id="acme") as sess:
        response = client.messages.create(...)
    print(f"Cost: ${sess.total_cost_usd:.4f}")
"""

from __future__ import annotations

import contextvars
import time
from typing import TYPE_CHECKING, Any

from opentelemetry import trace as otel_trace
from opentelemetry.trace import Span, Status, StatusCode

if TYPE_CHECKING:
    from claude_trace._config import TraceConfig

_TRACER_NAME = "claude-trace"

_SESSION_CONTEXT_VAR: contextvars.ContextVar["AgentSession | None"] = contextvars.ContextVar(
    "claude_trace_session", default=None
)


class AgentSession:
    """Tracks a complete agent conversation as an OTel root span.

    All API calls made within this context manager are attributed to this
    session, enabling per-session cost tracking, trace grouping, and metadata.
    """

    def __init__(
        self,
        name: str,
        customer_id: str | None,
        tags: list[str],
        metadata: dict[str, Any],
        config: "TraceConfig",
    ) -> None:
        self.name = name
        self.customer_id = customer_id
        self.tags = tags
        self.metadata = metadata
        self.config = config
        self.total_input_tokens: int = 0
        self.total_output_tokens: int = 0
        self.total_cache_read_tokens: int = 0
        self.total_cache_write_tokens: int = 0
        self.total_cost_usd: float = 0.0
        self.turn_count: int = 0
        self._span: Span | None = None
        self._token: contextvars.Token["AgentSession | None"] | None = None
        self._start_time: float = 0.0

    def __enter__(self) -> "AgentSession":
        tracer = otel_trace.get_tracer(_TRACER_NAME)
        self._span = tracer.start_span("claude.agent.session")
        self._span.set_attribute("claude.session.name", self.name)
        if self.customer_id and not self.config.sanitize:
            self._span.set_attribute("claude.session.customer_id", self.customer_id)
        if self.tags:
            self._span.set_attribute("claude.session.tags", ",".join(self.tags))
        for k, v in self.metadata.items():
            if isinstance(v, (str, int, float, bool)):
                self._span.set_attribute(f"claude.session.{k}", v)
        self._token = _SESSION_CONTEXT_VAR.set(self)
        self._start_time = time.monotonic()
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        if self._span is None:
            return
        self._span.set_attribute("claude.session.total_input_tokens", self.total_input_tokens)
        self._span.set_attribute("claude.session.total_output_tokens", self.total_output_tokens)
        self._span.set_attribute(
            "claude.session.total_cache_read_tokens", self.total_cache_read_tokens
        )
        self._span.set_attribute("claude.session.turn_count", self.turn_count)
        self._span.set_attribute(
            "claude.session.estimated_cost_usd", f"{self.total_cost_usd:.6f}"
        )
        if exc_type is not None:
            self._span.set_status(Status(StatusCode.ERROR, str(exc_val)))
        else:
            self._span.set_status(Status(StatusCode.OK))
        self._span.end()
        if self._token is not None:
            _SESSION_CONTEXT_VAR.reset(self._token)

    def _record_turn(
        self,
        input_tokens: int,
        output_tokens: int,
        cache_read: int,
        cache_write: int,
        cost_usd: float,
    ) -> None:
        """Called by the instrumentation layer after each API call."""
        self.total_input_tokens += input_tokens
        self.total_output_tokens += output_tokens
        self.total_cache_read_tokens += cache_read
        self.total_cache_write_tokens += cache_write
        self.total_cost_usd += cost_usd
        self.turn_count += 1


def current_session() -> "AgentSession | None":
    """Return the active AgentSession from context, or None."""
    return _SESSION_CONTEXT_VAR.get()

File: typescript/src/_session.ts
import { trace, context, SpanStatusCode } from '@opentelemetry/api';
import type { Span, Context } from '@opentelemetry/api';
import type { TraceConfig } from './_config';

const TRACER_NAME = 'claude-trace';

let _currentSession: AgentSession | undefined;

export class AgentSession {
  readonly name: string;
  readonly span: Span;
  private readonly _config: TraceConfig;
  private readonly _prevContext: Context;
  totalInputTokens = 0;
  totalOutputTokens = 0;
  totalCacheReadTokens = 0;
  totalCacheWriteTokens = 0;
  totalCostUsd = 0;
  turnCount = 0;

  constructor(
    name: string,
    opts: { customerId?: string; tags?: string[]; config: TraceConfig },
  ) {
    this.name = name;
    this._config = opts.config;
    this._prevContext = context.active();

    const tracer = trace.getTracer(TRACER_NAME);
    this.span = tracer.startSpan('claude.agent.session');
    this.span.setAttribute('claude.session.name', name);
    if (opts.customerId != null && !opts.config.sanitize) {
      this.span.setAttribute('claude.session.customer_id', opts.customerId);
    }
    if (opts.tags != null && opts.tags.length > 0) {
      this.span.setAttribute('claude.session.tags', opts.tags.join(','));
    }
  }

  /** @internal Called by the instrumentation layer after each API call. */
  _recordTurn(
    inputTokens: number,
    outputTokens: number,
    cacheRead: number,
    cacheWrite: number,
    costUsd: number,
  ): void {
    this.totalInputTokens += inputTokens;
    this.totalOutputTokens += outputTokens;
    this.totalCacheReadTokens += cacheRead;
    this.totalCacheWriteTokens += cacheWrite;
    this.totalCostUsd += costUsd;
    this.turnCount += 1;
  }

  end(error?: Error): void {
    this.span.setAttribute('claude.session.total_input_tokens', this.totalInputTokens);
    this.span.setAttribute('claude.session.total_output_tokens', this.totalOutputTokens);
    this.span.setAttribute('claude.session.turn_count', this.turnCount);
    this.span.setAttribute('claude.session.estimated_cost_usd', this.totalCostUsd.toFixed(6));
    if (error != null) {
      this.span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
    } else {
      this.span.setStatus({ code: SpanStatusCode.OK });
    }
    this.span.end();
    _currentSession = undefined;
  }
}

/** Return the active AgentSession, or undefined. */
export function currentSession(): AgentSession | undefined {
  return _currentSession;
}

/**
 * Run a callback within a named agent session.
 *
 * @example
 * ```typescript
 * const result = await session('billing-agent', { customerId: 'acme' }, async (sess) => {
 *   const response = await client.messages.create({...});
 *   return response;
 * });
 * ```
 */
export async function session<T>(
  name: string,
  opts: { customerId?: string; tags?: string[]; config?: TraceConfig },
  fn: (sess: AgentSession) => Promise<T>,
): Promise<T> {
  const { TraceConfig } = await import('./_config');
  const cfg = opts.config ?? TraceConfig.fromEnv();
  const sess = new AgentSession(name, { ...opts, config: cfg });
  _currentSession = sess;
  try {
    const result = await fn(sess);
    sess.end();
    return result;
  } catch (err) {
    sess.end(err instanceof Error ? err : new Error(String(err)));
    throw err;
  }
}
